tests/sanity.cpp

```diff
@@ -1508,6 +1508,102 @@ BEGIN_GDRCOPY_TEST(invalidation_fork_child_gdr_pin_parent_with_tokens)
 }
 END_GDRCOPY_TEST
 
+struct mt_test_info {
+    CUdeviceptr d_buf;
+    void *mapped_d_buf;
+    size_t size;
+    gdr_t g;
+    gdr_mh_t mh;
+    bool use_barrier;
+    pthread_barrier_t barrier;
+};
+
+void *thr_fun_setup(void *data)
+{
+    mt_test_info *pt = static_cast<mt_test_info*>(data);
+    ASSERT(pt);
+    ASSERT_EQ(gdr_pin_buffer(pt->g, pt->d_buf, pt->size, 0, 0, &pt->mh), 0);
+    ASSERT_NEQ(pt->mh, null_mh);
+    ASSERT_EQ(gdr_map(pt->g, pt->mh, &pt->mapped_d_buf, pt->size), 0);
```

Comment:
I guess we're not breaking the test down further where threadA calls pin_buffer but a different thread calls map. This should probably be guaranteed by calling thread by grabbing lock before pin and releasing lock after map succeeds or fails.

Reply:
if you think it might be useful, I can add a finer grained test

Reply:
From UCT viewpoint, [pin](https://github.com/openucx/ucx/blob/master/src/uct/cuda/gdr_copy/gdr_copy_md.c#L120) and [map](https://github.com/openucx/ucx/blob/master/src/uct/cuda/gdr_copy/gdr_copy_md.c#L126) are part of the same call [mem_reg](https://github.com/openucx/ucx/blob/master/src/uct/cuda/gdr_copy/gdr_copy_md.c#L106), so we're probably safe and we will likely add locks to ensure safe use of memory handle returned by gdr_map. If some other user of gdrcopy decides not to do this, then is it unsafe? For instance, what happens if threadA calls gdr_pin_buffer but the same memory handle is used by threadB and threadC concurrently?

Reply:
> we will likely add locks to ensure safe use of memory handle returned by gdr_map

why do you need locks?

>  For instance, what happens if threadA calls gdr_pin_buffer but the same memory handle is used by threadB and threadC concurrently?

for correctness, threadB and C would still need to wait for mem handle creation and mapping, right?
once the mem handle gets created and mapped, both B and C can use it, since it is a plain piece of virtual address space.

Reply:
> why do you need locks?

The aim was to ensure that two threads don't try to pin the same underlying cuda region concurrently. Is that considered safe? If so, then we don't have to use locks.

> for correctness, threadB and C would still need to wait for mem handle creation and mapping, right?

Agreed that there is a dependency on completion of pin_buffer.

Reply:
> why do you need locks?

Just had a thought about this. Should we add the case where main thread allocates gpu mem and two child threads try to run pin+map on the same allocation?

Something like this:
```C++
    {
        pthread_t tid[2];
        ASSERT_EQ(pthread_barrier_init(&t.barrier, NULL, 2), 0);
        t.use_barrier = true;
        print_dbg("spawning two children threads\n");
        ASSERT_EQ(pthread_create(&tid[0], NULL, thr_fun_combined, &t), 0);
        ASSERT_EQ(pthread_create(&tid[1], NULL, thr_fun_combined, &t), 0);
        ASSERT_EQ(pthread_join(tid[0], NULL), 0);
        ASSERT_EQ(pthread_join(tid[1], NULL), 0);
    }
```

Reply:
> Just had a thought about this. Should we add the case where main thread allocates gpu mem and two child threads try to run pin+map on the same allocation?

Good idea. Just pushed that subtest. PLMK

Reply:
Confirmed that the unit test and ucx-test still pass with the new changes. This confirms that it's not necessary to use locks to safely use pin buffer and map calls. 

Out of curiosity, I also checked to see if it's safe to call pin_buffer and map with different buffers and that also seems safe.